accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,375 @@
+
+```dart
+doMyOtherStuff(doMyStuff(something.doStuff()).doOtherStuff())
+```
+
+That code is so much less readable than
+
+```dart
+something.doStuff().doMyStuff().doOtherStuff().doMyOtherStuff()
+```
+
+So, primarily for readability reasons, static extension methods will allow you to add "extension methods", which are really just static functions, to existing types, and allow you to call those methods as if they were actual methods, using `.`.
+
+The extension methods are *static*, which means that we use the static type of an expression to figure out which method to call, and that also means that static extension methods are not virtual methods.
+
+The methods we allow you to add this way are normal methods, operators, getter and setters.
+
+## Declaring Static Extension Methods
+
+### Syntax
+
+A static extension of a type is declared using syntax like:
+
+```dart
+extension MyFancyList<T> on List<T> {
+  int get doubleLength => this.length * 2;
+  List<T> operator-() => this.reversed.toList();
+  List<List<T>> split(int at) => 
+      <List<T>>[this.sublist(0, at), this.sublist(at)];
+  List<T> mapToList<R>(R Function(T) convert) => this.map(convert).toList();
+}
+```
+
+More precisely, an extension declaration is a top-level declaration with a grammar similar to:
+
+```ebnf
+<extension> ::= 
+  `extension' <identifier> <typeParameters>? `on' <type> `?'? `{'
+     memberDeclaration*
+  `}'
+```
+
+Such a declaration introduces its *name* (the identifier) into the surrounding scope. The name does not denote a type, but it can be used to denote the extension itself in various places. The name can be hidden or shown in `import` or `export` declarations.
+
+The *type* can be any valid Dart type, but not a single type variable. It can refer to the type parameters of the extension. It can be followed by `?` which means that it allows `null` values.
+
+The member declarations can be any static or instance member declaration except for instance variables and constructors.
+
+### Scope
+
+Dart's static extension methods are *scoped*. They only apply to code that the extension itself is accessible to. Being in accessible means that using the extension's name must denote the extension. The extension is not in scope if another declaration with the same name shadows the extension, if the extension's library is not imported, if the library is imported and the extension is hidden, or the library is only imported with a prefix. In other words, if the extension had been a class, it is only in scope if using the name would denote the class&mdash; which will allow you to call static methods on the class, which is exactly what we are going to do.
```

Comment:
You can alwasy do scope override with import prefixes, you just have to do the *explicit* `quiver.CoolList(myList).foo()` override. You don't get the implicit suffix invocation when you override.

It is true that a conflict of name will prevent you from accessing the extension, just as it would prevent you from accessing a class. If you have two different classes names `CoolList`, then you can't call a static method on either.

We could choose to allow conflicts where all the names are extensions, as long as the member resolution gives a unique result, so if you have two `CoolList` extensions in scope, but only one of them has a `foo` member, then `[42].foo()` is not considered a conflict. Or if both have a `foo`, but one is more specific than the other, then it is not considered a conflict.
I am wary at allowing this because it would suggest that declaring those in the same scope should also be allowed (you can have them in the same scope, so why not declared them in the same scope?), and then people might think the following is a reasonable design:
```dart
extension CoolList<T> on List<T> { T foo() ... }
extension CoolList on List<int> { int foo() ...int specific code ... }
extension CoolList on List<String> { String foo() ...String specific code ... }
```
If library authors start doing that, then you *cannot* do an explicit override any more because that depends on explicit name resolution.

I think keeping the "at most one definition of a name in each scope" rule is important for language sanity.

That leaves us with the ability to *open* more than one extension with the same name in a specific scope, which means renaming.

We can introduce the "extension alias" you suggest. We could also go one step further and introduce a renaming operation directly in the import modifiers:
```dart
import "package:list/list.dart" show CoolList;
import "package:quiver/quiver.dart" show CoolList as QuiverList;
// or: import "package:quiver/quiver.dart" hide CoolList as QuiverList;
```
That would allow renaming any import. It's a trivial rename, all uses of the new name will refer to the imported declaration.
In the importing library, if the name `QuiverList` is in scope and not shadowed, then the quiver `CoolList` extension applies.

I'm sure people will have lots of fun with `import "dart:core" hide dynamic as Any;`. :smiling_imp:

Reply:
> Then again, I'm not sure why we will need to give more ways to avoid scope conflict for extensions than we do for other declarations.

My argument would be that prefixing other kinds of things doesn't change their affordances.  There is nothing you can do with a class `Foo` that you can't do with `foo.Foo`, as far as I know.  And similarly for functions, getters, setters, etc.  But with extensions, there is: prefixing it stops the implicit open of the extension in scope.  I'm really worried that if providing extensions as part of APIs becomes really common, then basically everyone will have to name mangle their extension names to be globally unique or else clients won't be able to use them as intended.  

I'd be fine with import based renaming.  I'd be fine with explicit renaming.  

Another approach would be to allow "opening" and extension explicitly.  So you would do:

```dart
import "package:list/list.dart" show CoolList;  // Bring CoolList into scope, but not CoolList.foo
open CoolList;  // Bring CoolList.foo into the extension method resolution scope.
```

Reply:
I have a scope override for opting in to an extension, but none for opting out. You need to remove the extension from the scope for that. That is probably unavoidable since we use the same syntax for everything, so there is no syntax we can write to do "the normal thing", because that is the syntax we are using already.

Reply:
We could use generalized type aliases to perform renaming, just like they can rename classes:

```dart
import `pkg:list` as list;
import `pkg:quiver` as quiver;

typedef CoolList<X> = list.CoolList<X>;
typedef QuiverList<X> = quiver.CoolList<X>;
```

These type aliases would be semantically isolated from other type aliases (which is good because it means that we don't promise being able to do anything new just because we're accessing a renamed extension), but it might also be useful because it can be used to pre-resolve some conflicts in a way that will work for the given application:

```dart
import `pkg:list` as list;
import `pkg:quiver` as quiver;

typedef CoolList<X> = list.CoolList<X>;
typedef QuiverList<X extends SomeQuiverishType> = quiver.CoolList<X>;
```

So if `quiver.CoolList` admits all type arguments, but this application will actually only use it when `X extends SomeQuiverishType` then we could use the above to ensure that there is no conflict with `list.CoolList` for any type argument which does not extend `SomeQuiverishType`.

Reply:
It feels really off to me to use typedef for renaming these since they... aren't types.  It feels really misleading and confusing.  If we want a declaration, maybe:

```dart
extension CoolList<X> = list.CoolList<X>;
```

?